faunadb-java-dsl/src/main/java/com/faunadb/client/types/Encoder.java

```diff
@@ -0,0 +1,205 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.errors.FaunaException;
+import com.faunadb.client.types.Properties.Property;
+import com.faunadb.client.types.Value.*;
+import com.faunadb.client.types.time.HighPrecisionTime;
+import org.joda.time.Instant;
+import org.joda.time.LocalDate;
+
+import java.lang.reflect.Array;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Stack;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Lists.newArrayListWithCapacity;
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+
+/**
+ * FaunaDB object to {@link Value} encoder.
+ */
+public class Encoder {
+    final static Stack<Object> stack = new Stack<>();
+    final static ConcurrentHashMap<Class<?>, Property[]> READ_PROPERTIES = new ConcurrentHashMap<>();
+    final static ConcurrentHashMap<Class<?>, Map<Enum<?>, StringV>> ENUMS = new ConcurrentHashMap<>();
+
+    /**
+     * Encode the specified object into a corresponding FaunaDB value.
+     *
+     * <pre>{@code
+     * encode(10) => new LongV(10)
+     * encode(3.14) => new DoubleV(3.14)
+     * encode(true) => BooleanV.TRUE
+     * encode(null) => NullV.NULL
+     * encode("a string") => new StringV("a string")
+     * encode(new int[] {1, 2}) => new ArrayV(Arrays.asList(1, 2))
+     * encode(Arrays.asList(1, 2)) => new ArrayV(Arrays.asList(1, 2))
+     * encode(new byte[] {1, 2}) => new ByteV(new byte[] {1, 2})
+     * encode(user) => new ObjectV(ImmutableMap.of("user_name", new StringV("john"), "password", new StringV("s3cr3t")))
+     * }</pre>
+     *
+     * @param obj Any instance of user defined classes, primitive values or any
+     *            generic collection like {@link java.util.List}, {@link java.util.Set} or {@link java.util.Map}
+     * @return A FaunaDB {@link Value} corresponding to the given argument
+     */
+    public static Value encode(Object obj) {
+        if (obj == null)
+            return Value.NullV.NULL;
+
+        if (Value.class.isInstance(obj))
+            return (Value) obj;
+
+        if (stack.contains(obj))
+            throw new FaunaException(format("Self reference loop detected for object `%s`", obj));
+
+        try {
+            stack.push(obj);
+
+            return encodeIntern(obj);
+        } finally {
+            stack.pop();
+        }
+    }
+
+    private static Value encodeIntern(Object obj) {
+        Class<?> clazz = obj.getClass();
+
+        if (clazz.isEnum())
+            return wrapEnum((Enum<?>) obj, clazz);
+
+        if (clazz == String.class)
```

Comment:
If we manage to use the `Codec` to decode data. We should try to use them here as well. Like the way we did for Scala driver.